### PR TITLE
fix tween color

### PR DESCRIPTION
**Pull request by zxx43**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

issue: https://github.com/cocos/cocos-engine/issues/13229

### CODE REVIEW

File: cocos/tween/tween-action.ts [hunk 1]
```diff
@@ -217,18 +275,51 @@ export class TweenAction extends ActionInterval {
...
+                } else if (prop.type && prop.type.get('vec4')) {
+                    prop.current.x = interpolation(start.x, end.x, prop.current.x, time);
+                    prop.current.y = interpolation(start.y, end.y, prop.current.y, time);
+                    prop.current.z = interpolation(start.z, end.z, prop.current.z, time);
+                    prop.current.w = interpolation(start.w, end.w, prop.current.w, time);
+                } else if (prop.type && prop.type.get('size')) {
+                    prop.current.width = interpolation(start.width, end.width, prop.current.width, time);
+                    prop.current.height = interpolation(start.height, end.height, prop.current.height, time);
+                } else if (prop.type && prop.type.get('quat')) {
+                    prop.current.x = interpolation(start.x, end.x, prop.current.x, time);
```
Comment (SantyWang) [inline cocos/tween/tween-action.ts]: The quaternion cannot be interpolated into four fields.

File: cocos/tween/tween-action.ts [hunk 2]
```diff
@@ -149,6 +150,7 @@ export class TweenAction extends ActionInterval {
 
             const prop = Object.create(null);
             prop.value = value;
+            prop.type = new Map<string, boolean>();
```
Comment (SantyWang) [inline cocos/tween/tween-action.ts]: no needs to use map? string seems better?

File: tests/tween/tween.test.ts [hunk 3]
```diff
@@ -38,4 +41,102 @@ test('destroySelf', function () {
...
+    const shadow = new Shadows();
+    const tweenact = tween(shadow).to(1, {shadowColor: Color.TRANSPARENT});
+    tweenact.start();
+    // @ts-expect-error access private property 
+    const action = tweenact._actions[0] as TweenAction;
+    // @ts-expect-error access private property 
+    const props = action._props;
+    for (const property in props) {
+        const prop = props[property];
+        expect(prop.current instanceof Color).toBeTruthy();
```
Comment (SantyWang) [inline tests/tween/tween.test.ts:66]: need to check the animation result.

File: cocos/tween/tween-action.ts [hunk 4]
```diff
@@ -217,18 +275,48 @@ export class TweenAction extends ActionInterval {
...
+                } else if (prop.type && prop.type === 'vec4') {
+                    prop.current.x = interpolation(start.x, end.x, prop.current.x, time);
+                    prop.current.y = interpolation(start.y, end.y, prop.current.y, time);
+                    prop.current.z = interpolation(start.z, end.z, prop.current.z, time);
+                    prop.current.w = interpolation(start.w, end.w, prop.current.w, time);
+                } else if (prop.type && prop.type === 'size') {
+                    prop.current.width = interpolation(start.width, end.width, prop.current.width, time);
+                    prop.current.height = interpolation(start.height, end.height, prop.current.height, time);
+                } else if (prop.type && prop.type === 'quat') {
+                    Quat.slerp(prop.current, start, end, time as number);
```
Comment (SantyWang) [inline cocos/tween/tween-action.ts:316]: Should output some warning if there is a custom interpolation method to tell that quat does not support custom interpolation method.

File: cocos/tween/tween-action.ts [hunk 5]
```diff
@@ -217,18 +275,48 @@ export class TweenAction extends ActionInterval {
...
+                } else {
+                    for (const k in start) {
+                        // if (value[k].easing) {
```
Comment (SantyWang) [inline cocos/tween/tween-action.ts]: Delete comment.

File: tests/tween/tween.test.ts [hunk 6]
```diff
@@ -38,4 +41,131 @@ test('destroySelf', function () {
...
+test('type color', function () {
+    const shadow = new Shadows();
+    shadow.shadowColor.set(200, 100, 50, 250);
+    const target = Color.TRANSPARENT;
+    const tweenact = tween(shadow).to(1, {shadowColor: target});
+    tweenact.start();
+
+    // @ts-expect-error access private property 
+    const action = tweenact._actions[0] as TweenAction;
+    action.update(1.0);
```
Comment (SantyWang) [inline tests/tween/tween.test.ts]: use game.step to update tween.

File: tests/tween/tween.test.ts [hunk 7]
```diff
@@ -38,4 +41,131 @@ test('destroySelf', function () {
...
     expect(onDestroy).toBeCalledTimes(1);
     director.unregisterSystem(sys);
+});
+
+test('type color', function () {
+    const shadow = new Shadows();
+    shadow.shadowColor.set(200, 100, 50, 250);
+    const target = Color.TRANSPARENT;
+    const tweenact = tween(shadow).to(1, {shadowColor: target});
+    tweenact.start();
```
Comment (SantyWang) [inline tests/tween/tween.test.ts:55]: Should test custom easing of all value types.

File: cocos/tween/tween-action.ts [hunk 8]
```diff
@@ -180,14 +183,69 @@ export class TweenAction extends ActionInterval {
...
+                if (_t instanceof Vec2) {
+                    if (prop.start == null) {
+                        prop.start = new Vec2(); prop.current = new Vec2(); prop.end = new Vec2();
+                    }
+                    prop.type = 'vec2';
+                } else if (_t instanceof Vec3) {
+                    if (prop.start == null) {
+                        prop.start = new Vec3(); prop.current = new Vec3(); prop.end = new Vec3();
+                    }
+                    prop.type = 'vec3';
```
Comment (SantyWang) [inline cocos/tween/tween-action.ts:238]: Why don't handle relative?

**Comment by dumganhar** [comment]
Submit another PR https://github.com/cocos/cocos-engine/pull/16977 here.
Close this one.

**Pull request closed without merging**